File: lib/gateway.js
const { URL } = require('url');

function parseGatewayHost(baseUrl) {
  const raw = String(baseUrl || '').trim();
  if (!raw) return '';
  try {
    const url = new URL(raw.includes('://') ? raw : `https://${raw}`);
    return String(url.hostname || '').trim().toLowerCase();
  } catch {
    try {
      return raw.split('/')[0].split('@').pop().split(':')[0].trim().toLowerCase();
    } catch {
      return '';
    }
  }
}

function providerLabelForHost(host) {
  const h = String(host || '').toLowerCase();
  if (!h) return '';
  if (h.includes('z.ai') || h.includes('bigmodel.cn')) return 'GLM';
  if (h.includes('openrouter')) return 'OpenRouter';
  if (h.includes('moonshot') || h.includes('kimi')) return 'Kimi';
  if (h.includes('deepseek')) return 'DeepSeek';
  return h;
}

function displayHostForHost(host) {
  const h = String(host || '').toLowerCase();
  if (!h) return '';
  if (h.includes('z.ai')) return 'z.ai';
  if (h.includes('bigmodel.cn')) return 'bigmodel.cn';
  return h.replace(/^api\./, '');
}

function gatewayInfo({ env = process.env, settings = {}, config = {} } = {}) {
  const settingsEnv = settings && typeof settings.env === 'object' && !Array.isArray(settings.env) ? settings.env : {};
  const baseUrl = String(env.ANTHROPIC_BASE_URL || settingsEnv.ANTHROPIC_BASE_URL || '').trim();
  const authToken = String(env.ANTHROPIC_AUTH_TOKEN || settingsEnv.ANTHROPIC_AUTH_TOKEN || '').trim();
  const host = parseGatewayHost(baseUrl);
  const awareness = !config || config.gateway_awareness !== false;
  const foreign = Boolean(awareness && host && host !== 'api.anthropic.com');
  const label = providerLabelForHost(host);
  const displayHost = displayHostForHost(host);
  return { baseUrl, authTokenPresent: Boolean(authToken), host, displayHost, label, foreign };
}

function isForeignGateway(baseUrl) {
  return gatewayInfo({ env: { ANTHROPIC_BASE_URL: baseUrl } }).foreign;
}

function gatewayBadgeText(info) {
  if (!info || !info.foreign) return '';
  const host = info.displayHost || info.host || '';
  const label = info.label || host;
  if (!host) return '';
  return label && label !== host ? `via ${host} (${label})` : `via ${host}`;
}

function gatewayNoteLabel(info) {
  if (!info || !info.foreign) return '';
  return info.label || info.displayHost || info.host || '';
}

module.exports = {
  parseGatewayHost,
  providerLabelForHost,
  displayHostForHost,
  gatewayInfo,
  isForeignGateway,
  gatewayBadgeText,
  gatewayNoteLabel,
};

File: lib/gateway.py
"""Gateway/provider detection helpers for the statusline engines."""

from __future__ import annotations

import os
from urllib.parse import urlparse


def parse_gateway_host(base_url: str | None) -> str:
    raw = str(base_url or "").strip()
    if not raw:
        return ""
    try:
        parsed = urlparse(raw if "://" in raw else f"https://{raw}")
        return (parsed.hostname or "").strip().lower()
    except Exception:
        try:
            return raw.split("/", 1)[0].split("@")[-1].split(":", 1)[0].strip().lower()
        except Exception:
            return ""


def provider_label_for_host(host: str | None) -> str:
    h = str(host or "").lower()
    if not h:
        return ""
    if "z.ai" in h or "bigmodel.cn" in h:
        return "GLM"
    if "openrouter" in h:
        return "OpenRouter"
    if "moonshot" in h or "kimi" in h:
        return "Kimi"
    if "deepseek" in h:
        return "DeepSeek"
    return h


def display_host_for_host(host: str | None) -> str:
    h = str(host or "").lower()
    if not h:
        return ""
    if "z.ai" in h:
        return "z.ai"
    if "bigmodel.cn" in h:
        return "bigmodel.cn"
    if h.startswith("api."):
        return h[4:]
    return h


def gateway_info(env=None, settings=None, config=None) -> dict:
    if env is None:
        env = os.environ
    settings = settings if isinstance(settings, dict) else {}
    config = config if isinstance(config, dict) else {}
    settings_env = settings.get("env") if isinstance(settings.get("env"), dict) else {}
    base_url = str(env.get("ANTHROPIC_BASE_URL") or settings_env.get("ANTHROPIC_BASE_URL") or "").strip()
    auth_token = str(env.get("ANTHROPIC_AUTH_TOKEN") or settings_env.get("ANTHROPIC_AUTH_TOKEN") or "").strip()
    host = parse_gateway_host(base_url)
    awareness = config.get("gateway_awareness") is not False
    foreign = bool(awareness and host and host != "api.anthropic.com")
    label = provider_label_for_host(host)
    display_host = display_host_for_host(host)
    return {
        "base_url": base_url,
        "auth_token_present": bool(auth_token),
        "host": host,
        "display_host": display_host,
        "label": label,
        "foreign": foreign,
    }


def is_foreign_gateway(base_url: str | None) -> bool:
    return gateway_info(env={"ANTHROPIC_BASE_URL": base_url})["foreign"]


def gateway_badge_text(info: dict | None) -> str:
    if not isinstance(info, dict) or not info.get("foreign"):
        return ""
    host = info.get("display_host") or info.get("host") or ""
    label = info.get("label") or host
    if not host:
        return ""
    if label and label != host:
        return f"via {host} ({label})"
    return f"via {host}"


def gateway_note_label(info: dict | None) -> str:
    if not isinstance(info, dict) or not info.get("foreign"):
        return ""
    return info.get("label") or info.get("display_host") or info.get("host") or ""
